### ISSUE TITLE
Lazy load icons on website

### ISSUE THREAD

**Issue by ericcornelissen** [simple-icons/simple-icons-website#37]
### Kind of issue 
  - [x] Improvement
  - [ ] Bug
  - [ ] Other, namely: 

### This issue concerns 
  - [ ] The NPM Package
  - [ ] The Packagist Package
  - [x] The website
  - [ ] Other, namely: 

### Description

With [over a 1000 icons](https://github.com/simple-icons/simple-icons/pull/2365) I think it might be a good idea to implement the lazy loading of icons on the website. That is, when the page is first loaded we load a limited amount of icons (say 250) and when the user scrolls down we (gradually?) load more icons.

Some things to keep in mind/that should be tested thoroughly:

- How will this work with the different sorting methods (colour vs. alphabetical)
- How will this work with searching
- More...?

**Comment by ericcornelissen** [comment]
> we (gradually?) load more icons.

I do think we should do it gradually. That is, I don't think we should automatically start loading all icons in the background even if the user does not interact with the page - this seems like a waste of internet connection to me.


> How will this work with the different sorting methods (colour vs. alphabetical)

I think this is one of the most difficult to tackle. Especially w.r.t. the fact that we don't know which of the two the user has selected before the page is loaded. The striaghtforward solution would be to not pre-load any SVGs, and start loading SVGs only once the ordering is known. This may work, but I think it should be tested how this affects the loading experience of the website.

Alternative suggestions are welcome.


> How will this work with searching

This should be fairly straightforward, as we can just load the SVGs that match the search query that where not previously loaded. We can easily keep list of all brand names in memory for searching (even with https://github.com/simple-icons/simple-icons/issues/4797), search in that and load the icons that are needed on demand.

A further optimization here would be to also load/show only the first X icons matching a search term, though depending on the value of X, I'm not sure if that currently has any effect.


> More 1: Supporting non-JavaScript users

Following the lead of https://github.com/simple-icons/simple-icons-website/issues/34 and https://github.com/simple-icons/simple-icons-website/issues/35, a separate page could be generated with all icons preloaded (that is, the way the website works now). This version of the website can be linked to at the end of icons being loaded by default.

**Comment by ayyash** [comment]
Not sure about this lazy loading, the svgs are simple text, and the page loads fast enough, why do you want to lazy load? 

>I do think we should do it gradually. That is, I don't think we should automatically start loading all icons in the background even if the user does not interact with the page - this seems like a waste of internet connection to me.

a single scroll on a facebook page consumes much more internet than that, just saying :)

**Comment by ericcornelissen** [comment]
As per @JoelEllis' [suggestion](https://github.com/simple-icons/simple-icons/discussions/4865#discussioncomment-491081) we could choose to use `loading="lazy"` on `<img>` tags for the SVGs. Performance-wise this is a very good option as it reduces the load time by 50% on my machine (from ~5s to ~2.5s).

The main drawback is that [the `loading` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-loading) is still experimental, and not supported when JavaScript is disabled. If the `loading` attribute is not used, the website becomes "usable" after ~5s, but for me it can take anywhere from ~10s to ~15s to load all icons which is quite bad. Additionally, "usable" is in quotation marks because it responsive significantly slower to inputs such as hovering and scrolling.

Additionally, when using `<img>` we need to find a way to still support copying the SVG. This is doable, but just not as straightforward as it currently is.

**Comment by service-paradis** [comment]
`lazy="load"` is a pretty good solution. According to https://caniuse.com/loading-lazy-attr, it would work for about ~75% of users, which is not bad.
For the others, adding the `lazy` attribute wont break. It just wont work.

If we want to support more browsers, we could also use IntersectionObserver to lazy load SVGs when `lazy` attribute is not supported.

**Comment by ericcornelissen** [comment]
> `lazy="load"` is a pretty good solution. According to https://caniuse.com/loading-lazy-attr, it would work for about ~75% of users, which is not bad.
> For the others, adding the `lazy` attribute wont break. It just wont work.

Fair enough, if there is enough support for this I'm willing to make that trade-off.

 
> If we want to support more browsers, we could also use IntersectionObserver to lazy load SVGs when `lazy` attribute is not supported.

Not sure how that would work, as the browser will automatically start loading the SVGs. Unless the default is "don't load all icons" a solution like that won't work, I think.

**Comment by ericcornelissen** [comment]
I will be closing this for now with #55 merged. That does not mean we do not want to look into further lazy loading options in the future! If you have ideas feel free to share them here or open a new issue :slightly_smiling_face:
